From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matthias Schiffer <matthias.schiffer@ew.tq-group.com>
Date: Tue, 28 May 2024 16:31:37 +0200
Subject: [PATCH] mfd: tqmx86: configure board-specific GPIO platform data

Provide configuration to allow all GPIOs and interrupts to work on
TQMx86 SMARC modules. This will update the default direction of GPIOs 4
and 5 on SMARC modules to actually match the board documentation and
labels on the PCB.

Unknown boards are deliberately not supported (by not adding the GPIO
platform device), so future TQMx86 boards will not use incorrect
configuration on current kernels.

Signed-off-by: Matthias Schiffer <matthias.schiffer@ew.tq-group.com>
Upstream-Status: Pending
---
 drivers/mfd/tqmx86.c | 108 ++++++++++++++++++++++++++++++++++++-------
 1 file changed, 91 insertions(+), 17 deletions(-)

diff --git a/drivers/mfd/tqmx86.c b/drivers/mfd/tqmx86.c
index f087c66a7a202c7e986fcf7dd2e573c637fe32a9..cad2f835a1c08af24b226e159e17ce7216fded96 100644
--- a/drivers/mfd/tqmx86.c
+++ b/drivers/mfd/tqmx86.c
@@ -13,6 +13,7 @@
 #include <linux/io.h>
 #include <linux/mfd/core.h>
 #include <linux/module.h>
+#include <linux/platform_data/gpio-tqmx86.h>
 #include <linux/platform_data/i2c-machxo2.h>
 #include <linux/platform_data/i2c-ocores.h>
 #include <linux/platform_device.h>
@@ -98,19 +99,42 @@ static struct i2c_board_info tqmx86_i2c1_devices[] = {
 	},
 };
 
-static const struct mfd_cell tqmx86_devs[] = {
-	{
-		.name = "tqmx86-wdt",
-		.resources = tqmx_watchdog_resources,
-		.num_resources = ARRAY_SIZE(tqmx_watchdog_resources),
-		.ignore_resource_conflicts = true,
-	},
-	{
-		.name = "tqmx86-gpio",
-		.resources = tqmx_gpio_resources,
-		.num_resources = ARRAY_SIZE(tqmx_gpio_resources),
-		.ignore_resource_conflicts = true,
-	},
+static const struct mfd_cell tqmx86_wdt_dev = {
+	.name = "tqmx86-wdt",
+	.resources = tqmx_watchdog_resources,
+	.num_resources = ARRAY_SIZE(tqmx_watchdog_resources),
+	.ignore_resource_conflicts = true,
+};
+
+static const struct tqmx86_gpio_platform_data gpio_platform_data_come = {
+	.ngpio = 8,
+	.irq_valid_mask = {0x00f0}, /* GPIOs 0-3 have no IRQ support */
+	.default_dir_mask = {0x00f0}, /* 4 outputs, 4 inputs */
+};
+
+static const struct tqmx86_gpio_platform_data gpio_platform_data_smarc20 = {
+	.ngpio = 12,
+	.irq_valid_mask = {0x0fff}, /* IRQ support on all GPIOs */
+	.default_dir_mask = {0x0fc0}, /* 6 outputs, 6 inputs */
+};
+
+static const struct tqmx86_gpio_platform_data gpio_platform_data_smarc21 = {
+	.ngpio = 14,
+	.irq_valid_mask = {0x0fff}, /* GPIOs 12-13 have no IRQ support */
+	.default_dir_mask = {0x3fc0}, /* 6 outputs, 8 inputs */
+};
+
+static const struct tqmx86_gpio_platform_data gpio_platform_data_smarc21_full = {
+	.ngpio = 14,
+	.irq_valid_mask = {0x3fff}, /* IRQ support on all GPIOs */
+	.default_dir_mask = {0x3fc0}, /* 6 outputs, 8 inputs */
+};
+
+static struct mfd_cell tqmx86_gpio_dev = {
+	.name = "tqmx86-gpio",
+	.resources = tqmx_gpio_resources,
+	.num_resources = ARRAY_SIZE(tqmx_gpio_resources),
+	.ignore_resource_conflicts = true,
 };
 
 static const char *tqmx86_board_id_to_name(u8 board_id, u8 sauc)
@@ -191,6 +215,40 @@ static int tqmx86_board_id_to_clk_rate(struct device *dev, u8 board_id)
 	}
 }
 
+static const struct tqmx86_gpio_platform_data *
+tqmx86_board_id_to_gpio_pdata(u8 board_id, u8 sauc, u8 pld_rev)
+{
+	switch (board_id) {
+	case TQMX86_REG_BOARD_ID_E38M:
+	case TQMX86_REG_BOARD_ID_50UC:
+	case TQMX86_REG_BOARD_ID_E38C:
+	case TQMX86_REG_BOARD_ID_60EB:
+	case TQMX86_REG_BOARD_ID_E39C1:
+	case TQMX86_REG_BOARD_ID_E39C2:
+	case TQMX86_REG_BOARD_ID_70EB:
+	case TQMX86_REG_BOARD_ID_80UC:
+	case TQMX86_REG_BOARD_ID_120UC:
+	case TQMX86_REG_BOARD_ID_110EB:
+	case TQMX86_REG_BOARD_ID_E40M:
+	case TQMX86_REG_BOARD_ID_E40C1:
+	case TQMX86_REG_BOARD_ID_E40C2:
+	case TQMX86_REG_BOARD_ID_130UC:
+		return &gpio_platform_data_come;
+	case TQMX86_REG_BOARD_ID_E39MS:
+		return (sauc == 0xff)
+			? &gpio_platform_data_come
+			: &gpio_platform_data_smarc20;
+	case TQMX86_REG_BOARD_ID_E40S:
+		return (pld_rev > 1)
+			? &gpio_platform_data_smarc21_full
+			: &gpio_platform_data_smarc21;
+	case TQMX86_REG_BOARD_ID_E41S:
+		return &gpio_platform_data_smarc21_full;
+	default:
+		return NULL;
+	}
+}
+
 static int tqmx86_setup_irq(struct device *dev, const char *label, u8 irq,
 			    void __iomem *io_base, u8 reg_shift)
 {
@@ -316,7 +374,8 @@ static int tqmx86_detect_i2c(struct device *dev, const char *label,
 
 static int tqmx86_probe(struct platform_device *pdev)
 {
-	u8 board_id, sauc, rev;
+	const struct tqmx86_gpio_platform_data *gpio_pdata;
+	u8 board_id, sauc, rev, pcb_rev, pld_rev;
 	struct device *dev = &pdev->dev;
 	const char *board_name;
 	void __iomem *io_base;
@@ -330,10 +389,12 @@ static int tqmx86_probe(struct platform_device *pdev)
 	sauc = ioread8(io_base + TQMX86_REG_SAUC);
 	board_name = tqmx86_board_id_to_name(board_id, sauc);
 	rev = ioread8(io_base + TQMX86_REG_BOARD_REV);
+	pcb_rev = rev >> 4;
+	pld_rev = rev & 0xf;
 
 	dev_info(dev,
 		 "Found %s - Board ID %d, PCB Revision %d, PLD Revision %d\n",
-		 board_name, board_id, rev >> 4, rev & 0xf);
+		 board_name, board_id, pcb_rev, pld_rev);
 
 	if (gpio_irq) {
 		err = tqmx86_setup_irq(dev, "GPIO", gpio_irq, io_base,
@@ -355,9 +416,22 @@ static int tqmx86_probe(struct platform_device *pdev)
 	if (err)
 		return err;
 
+	gpio_pdata = tqmx86_board_id_to_gpio_pdata(board_id, sauc, pld_rev);
+	if (gpio_pdata) {
+		tqmx86_gpio_dev.platform_data = (void *)gpio_pdata;
+		tqmx86_gpio_dev.pdata_size = sizeof(*gpio_pdata);
+
+		err = devm_mfd_add_devices(dev, PLATFORM_DEVID_NONE,
+					   &tqmx86_gpio_dev, 1,
+					   NULL, 0, NULL);
+		if (err)
+			return err;
+	} else {
+		dev_warn(dev, "No GPIO support available\n");
+	}
+
 	return devm_mfd_add_devices(dev, PLATFORM_DEVID_NONE,
-				    tqmx86_devs,
-				    ARRAY_SIZE(tqmx86_devs),
+				    &tqmx86_wdt_dev, 1,
 				    NULL, 0, NULL);
 }
 
